[WIP] Implement `req` and `rep` atop PR #5

skrap opened the pull request:
This is a work in progress PR.

I'm working on porting the two remaining socket types, `REQ` and `REP`, to std::future, atop the changes made in PR #5.

I'm making this a PR now because I have a few questions:

- [x]  Idea: Not using `impl_wrapper_split` for REQ/REP:
These socket types must only be used in non-interleaved send/recv pairs, and it seemed to me that users will always want to keep a single object around to coordinate that fact.  Splitting the socket seemed fraught with peril.

- [x] Ok? Example is pretty liberal with unwrap() and devoid of error handling.

- [x] Error handling?
Speaking of error handling, `REQ` and `REP` have this interesting property where they must alternate between `recv` and `send` calls.  Naïvely, I would expect a `send` to a just-bound `REP` socket to `await` into a Err(EACCESS) as it does in the non-async `zmq` crate.  I wrote some (disabled) unit tests in reply.rs and request.rs which try to reproduce the non-async error.  However the code just hangs, and I haven't yet figured out what is being waited upon.

- [x] `REP` socket naming:
`REP` sockets seem to be named "reply" sockets in the zmq [guide](http://zguide.zeromq.org/page:all#Ask-and-Ye-Shall-Receive).  I have renamed the tmq APIs around this socket type to "reply" rather than "response"/"respond".  Is that ok?

skrap commented:
CC @Kobzol

Kobzol commented:
This general Sink/Stream implementation will do fine I think. But since we're doing a lot of changes already, what about trying to make the API in such a way that it is impossible to use it in a wrong way?

Rep and Req do not really represent a Stream nor a Sink IMO, since they are strictly interlocked, you cannot consume them in a loop as a normal Stream for example.

What if they had async methods, which would return a new structure allowing the other operation (send, recv), or the original structure if there was an error?

Something like this:
```rust
struct ReqSender;
impl ReqSender {
 pub async fn send(...) -> Result<ReqReceiver, (Error, ReqSender)>;
}

struct ReqReceiver;
impl ReqReceiver {
  pub async fn recv(...) -> Result<(ReqSender, Multipart), (Error, ReqReceiver)>;
}
```

Maybe it will be too unergonomic to use, but I think that it might be worth a try. What do you think?

skrap commented:
I considered doing that, yes, but I was worried about the ergonomics of unpacking the result as well as having an API for these socket types which was quite different.  However, given that you also had the same thought about a self-consuming `send` and `recv`, I will implement it and see how it feels to actually use it.

Kobzol commented:
Well REP/REQ are fundamentally different sockets, so I think that it's fine if we give them fundamentally different usage and semantics :) Let's see how it turns out.

skrap commented on examples/request.rs:
```diff
@@ -1,50 +1,34 @@
+
+        info!("Request: {:?}", &message);
+        let mut multipart = Multipart::from(vec![zmq::Message::from(message.as_bytes())]);
+        let recv_sock = send_sock.send(&mut multipart).await?;
```
@Kobzol ergonomics for a split sender-receiver object cluster seem pretty good to me.  Thoughts?  Also see `examples/reply.rs` for the opposite side of the loop.

Kobzol replied:
I like the interface! `examples/reply.rs` seems to be missing though?

skrap commented:
Still WIP: I have to figure out whether `send_sock.send(&mut multipart)` should actually be `send_sock.send(multipart)`, or maybe `send_sock.send(&multipart)`.

Kobzol commented:
I think that it should consume the message, since ZMQ takes ownership of the data pointed to by the multipart when it send them AFAIK. I know that it's not strictly equivalent, but I think that it might be much simpler (in terms of implementation) to implement this using `async/await`, what do you think?

https://github.com/Kobzol/tmq/commit/cc680c20d2550243486add72641e2a4ffd6110f3#diff-23b2d5d1c08fc79ec4af55ee06f0da8cR47

skrap commented:
Added missing reply example, and switched to a message-consuming `send`.

One thing to consider is how errors are handled.  The `send` function consumes the socket, but a client interested in error recovery might want a way to resurrect the socket in order to retry the `send`.  Towards this end, the `send` method returns a Result<> with a special error type which wraps a `TmqError` and a new RequestSender object, which contains the ZmqPoller and underlying zmq socket from the original `send` request.   This `SendError` can be `Into`-converted to a TmqError, so it can be used with `await?` or other try-operator expressions, if you want to just ignore recovery and drop the socket, which is what most implementations probably want.

The same technique is used for the Err() case of `recv()`.

I did this for ergonomics.  It means that the typical code which just wants to implement recovery via a `?`-return and creating a new socket will "just work".  It avoids most of the awkward destructuring of the result Err case.  However it does mean two new classes in the API.  I'm interested in feedback on the approach.

Kobzol commented:
For normal sockets, I think that the special error would be necessary. For ZMQ sockets, I'm not sure about the semantics - i.e. are ZMQ sockets truly "recoverable"? For example let's take a look at `zmq_send` (http://api.zeromq.org/3-2:zmq-send) and its possible errors:

- EAGAIN - handled by the async layer, won't be propagated
- ENOTSUP - wrong usage, shouldn't happen due to safe Rust API
- EFSM - wrong usage, shouldn't happen due to safe req/rep API
- ETERM - context terminated, shouldn't happen due to reference counting
- ENOTSOCK - invalid socket, shouldn't happen due to safe Rust API
- EINTR - not sure about this one, I hope that `zmq` deals with it inside, since I surely don't deal with it in the async wrapper :smile: 

So from this perspective (and from my current ZMQ usage so far), a ZMQ error is an unrecoverable event (and frankly shouldn't even happen often, as the sockets can recover from disconnects and various other things). But since you mentioned that you provide the easy `From` conversion, it shouldn't really matter, because with that the common unrecoverable path would be simple to use, so I think that it's fine.

What I would like to see more is the rewrite to async/await, since the code is just much simpler. What do you think about that?

skrap commented:
> What I would like to see more is the rewrite to async/await, since the code is just much simpler. What do you think about that?

Sorry, I am not understanding your meaning here.  What rewrite would you like to see?

Kobzol commented:
Remove the manual `RequestSend`/`impl Future for RequestSend` struct/future impl and just write it with async/await (that's the link that I sent before :-) https://github.com/Kobzol/tmq/commit/cc680c20d2550243486add72641e2a4ffd6110f3#diff-23b2d5d1c08fc79ec4af55ee06f0da8cR47 ).

skrap commented:
ohhh i didn't know about `futures::future::poll_fn`.  That's perfect.  Sorry I missed your earlier patch!

skrap commented:
Ok, I've researched failure handling in req/rep sockets, and it seems everyone just kills the socket and reconnects with a new one.

Sayeth the [guide](http://zguide.zeromq.org/page:all#Client-Side-Reliability-Lazy-Pirate-Pattern):
> If you try to use a REQ socket in anything other than a strict send/receive fashion, you'll get an error (technically, the REQ socket implements a small finite-state machine to enforce the send/receive ping-pong, and so the error code is called "EFSM"). This is slightly annoying when we want to use REQ in a pirate pattern, because we may send several requests before getting a reply.
> The pretty good brute force solution is to close and reopen the REQ socket after an error:

Also, from [Martin Sustrik](https://news.ycombinator.com/item?id=4161529):
> 1. As for lost requests, the right thing to do is to timeout the request and re-send the request if the response haven't been received. This should be done inside 0MQ/XS, however, it's pretty easy to do in the application so nobody so far felt incovenienced by it so much as to implement the feature.

From the context of the comment I think he means "connect a new socket and use it to re-send the request".

Given that, plus the evidence from error codes which you have shown above, it seems like one might not actually need to get at the socket again after a failure.  The code certainly does get smaller.

I do wonder if we shouldn't include an example of async plus timeout, since it's important to handle timeouts in ZMQ for several patterns, this one included.

Kobzol commented:
Examples with timeout would be nice. It's a bit tricky for sends though, if you get a timeout during send and you want to send anything else on the same socket, you have to `await` the original `send`, otherwise it's possible for the previous and following send to mix-up. For receives it should be fine, since ZMQ guarantees that a multipart will either be received at once or not at all.

skrap commented:
@Kobzol Well, with the changes I've put in place to drop the socket on errors, re-sending with the same socket isn't really an issue! 😄

Kobzol commented:
Haha, that's right! :) However it still needs to hold for the other socket types. But that's not relevant here, I just realized it when thinking about the timeout.

skrap commented:
Aha, this is interesting, and I think it opens the door to avoiding that ordering issue you mentioned, @Kobzol :

> On Thu, Apr 28, 2011 at 3:12 AM, MinRK wrote:
>
>>When sending or receiving multipart messages with NOBLOCK, is it possible for zmq_send or zmq_recv to raise EAGAIN on any part other than the first?
>
> In principle, no. If this happens it's a bug.
> -Pieter

from https://lists.zeromq.org/pipermail/zeromq-dev/2011-April/010697.html

So, we just have to make sure that when a sending future successfully sends the first part, it then doesn't yield until it's sent everything.  According to that post (and Pieter was one of the zmq authors) it should not block, so that should be fine, I think?  (I have not verified that tmq does the right thing at the moment, though.)

This matches what I found by a brief libzmq code inspection as well.

Kobzol commented:
Cool, I thought that it's only true for receiving, if it's also valid for sending, then it's fine. In principle tmq can abort a send in the middle, but only if a send returns EAGAIN, which should not happen as per the linked quote.

skrap commented:
See also https://github.com/zeromq/libzmq/issues/2269#issuecomment-269699914, which suggests that EAGAIN on non-first multipart messages should be treated as a full-message failure.

skrap commented:
I opened https://github.com/zeromq/libzmq/issues/3810 to see if we could get some clarity added to the docs.

cetra3 commented:
@skrap: As this stuff is rolled into the other PR, did you want to close this one off?

skrap commented:
Yep, closing.  See #5.
